File: app/src/main/java/com/walhalla/appextractor/model/PackageMeta.java
package com.walhalla.appextractor.model;

import android.content.ContentResolver;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;


public class PackageMeta implements Parcelable {

    public String packageName;
    public String label;
    public boolean hasSplits;
    public boolean hasPinning;
    public boolean isSystemApp;
    public long versionCode;
    public String versionName;
    public Uri iconUri;
    public long installTime;
    public long updateTime;

    public String size;
    public String sourceDir;

    public long firstInstallTime;
    public long lastUpdateTime;
    public int category;

    //Extended
    public boolean isGranted;

    public PackageMeta(String packageName, String label) {
        this.packageName = packageName;
        this.label = label;
    }


    public static class Builder {
        private final PackageMeta mPackageMeta;

        public Builder(String packageName) {
            mPackageMeta = new PackageMeta(packageName, "?");
        }

        public Builder setLabel(String label) {
            mPackageMeta.label = label;
            return this;
        }

        public Builder setHasSplits(boolean hasSplits) {
            mPackageMeta.hasSplits = hasSplits;
            return this;
        }

        public Builder setIsSystemApp(boolean isSystemApp) {
            mPackageMeta.isSystemApp = isSystemApp;
            return this;
        }

        public Builder setVersionCode(long versionCode) {
            mPackageMeta.versionCode = versionCode;
            return this;
        }

        public Builder setVersionName(String versionName) {
            mPackageMeta.versionName = versionName;
            return this;
        }

        public Builder setIcon(int iconResId) {
            if (iconResId == 0) {
                mPackageMeta.iconUri = null;
                return this;
            }

            mPackageMeta.iconUri = new Uri.Builder()
                    .scheme(ContentResolver.SCHEME_ANDROID_RESOURCE)
                    .authority(mPackageMeta.packageName)
                    .path(String.valueOf(iconResId))
                    .build();

            return this;
        }

        public Builder setIconUri(Uri iconUri) {
            mPackageMeta.iconUri = iconUri;
            return this;
        }


        public Builder setInstallTime(long installTime) {
            mPackageMeta.installTime = installTime;
            return this;
        }

        public Builder setUpdateTime(long updateTime) {
            mPackageMeta.updateTime = updateTime;
            return this;
        }

        public PackageMeta build() {
            return mPackageMeta;
        }

        public Builder setHasPining(boolean pining) {
            mPackageMeta.hasPinning = pining;
            return this;
        }
    }


//    public static PackageMeta forPackage(Context context, String packageName) {
//        try {
//            PackageManager pm = context.getPackageManager();
//
//            ApplicationInfo applicationInfo = pm.getApplicationInfo(packageName, 0);
//            PackageInfo packageInfo = pm.getPackageInfo(packageName, 0);
//
//            return new PackageMeta.Builder(applicationInfo.packageName)
//                    .setLabel(applicationInfo.loadLabel(pm).toString())
//                    .setHasSplits(applicationInfo.splitPublicSourceDirs != null && applicationInfo.splitPublicSourceDirs.length > 0)
//                    .setIsSystemApp((applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0)
//                    .setVersionCode(Utils.apiIsAtLeast(Build.VERSION_CODES.P) ? packageInfo.getLongVersionCode() : packageInfo.versionCode)
//                    .setVersionName(packageInfo.versionName)
//                    .setIcon(applicationInfo.icon)
//                    .setInstallTime(packageInfo.firstInstallTime)
//                    .setUpdateTime(packageInfo.lastUpdateTime)
//                    .build();
//
//        } catch (PackageManager.NameNotFoundException e) {
//            return null;
//        }
//    }

    protected PackageMeta(Parcel in) {
        packageName = in.readString();
        label = in.readString();
        hasSplits = in.readByte() != 0;
        isSystemApp = in.readByte() != 0;
        versionCode = in.readLong();
        versionName = in.readString();
        iconUri = in.readParcelable(Uri.class.getClassLoader());
        installTime = in.readLong();
        updateTime = in.readLong();
        size = in.readString();
        sourceDir = in.readString();
        firstInstallTime = in.readLong();
        lastUpdateTime = in.readLong();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(packageName);
        dest.writeString(label);
        dest.writeByte((byte) (hasSplits ? 1 : 0));
        dest.writeByte((byte) (isSystemApp ? 1 : 0));
        dest.writeLong(versionCode);
        dest.writeString(versionName);
        dest.writeParcelable(iconUri, flags);
        dest.writeLong(installTime);
        dest.writeLong(updateTime);
        dest.writeString(size);
        dest.writeString(sourceDir);
        dest.writeLong(firstInstallTime);
        dest.writeLong(lastUpdateTime);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<PackageMeta> CREATOR = new Creator<PackageMeta>() {
        @Override
        public PackageMeta createFromParcel(Parcel in) {
            return new PackageMeta(in);
        }

        @Override
        public PackageMeta[] newArray(int size) {
            return new PackageMeta[size];
        }
    };
}

File: extractor/src/main/java/com/walhalla/appextractor/model/PackageMeta.kt
package com.walhalla.appextractor.model

import android.content.ContentResolver
import android.net.Uri
import android.os.Parcel
import android.os.Parcelable


class PackageMeta : Parcelable {

    var packageName: String

    var label: String?
    var hasSplits: Boolean = false
    var hasPinning: Boolean = false
    var isSystemApp: Boolean = false
    var versionCode: Long = 0
    var versionName: String? = null

    var iconUri: Uri? = null
    var installTime: Long = 0
    var updateTime: Long = 0


    var size: String? = null

    var sourceDir: String? = null


    var firstInstallTime: Long = 0

    var lastUpdateTime: Long = 0

    var category: Int = 0

    //Extended

    var isGranted: Boolean = false

    constructor(packageName: String, label: String?) {
        this.packageName = packageName
        this.label = label
    }


    class Builder(packageName: String) {
        private val mPackageMeta = PackageMeta(packageName, "?")

        fun setLabel(label: String?): Builder {
            mPackageMeta.label = label
            return this
        }

        fun setHasSplits(hasSplits: Boolean): Builder {
            mPackageMeta.hasSplits = hasSplits
            return this
        }

        fun setIsSystemApp(isSystemApp: Boolean): Builder {
            mPackageMeta.isSystemApp = isSystemApp
            return this
        }

        fun setVersionCode(versionCode: Long): Builder {
            mPackageMeta.versionCode = versionCode
            return this
        }

        fun setVersionName(versionName: String?): Builder {
            mPackageMeta.versionName = versionName
            return this
        }

        fun setIcon(iconResId: Int): Builder {
            if (iconResId == 0) {
                mPackageMeta.iconUri = null
                return this
            }

            mPackageMeta.iconUri = Uri.Builder()
                .scheme(ContentResolver.SCHEME_ANDROID_RESOURCE)
                .authority(mPackageMeta.packageName)
                .path(iconResId.toString())
                .build()

            return this
        }

        fun setIconUri(iconUri: Uri?): Builder {
            mPackageMeta.iconUri = iconUri
            return this
        }


        fun setInstallTime(installTime: Long): Builder {
            mPackageMeta.installTime = installTime
            return this
        }

        fun setUpdateTime(updateTime: Long): Builder {
            mPackageMeta.updateTime = updateTime
            return this
        }

        fun build(): PackageMeta {
            return mPackageMeta
        }

        fun setHasPining(pining: Boolean): Builder {
            mPackageMeta.hasPinning = pining
            return this
        }
    }


    //    public static PackageMeta forPackage(Context context, String packageName) {
    //        try {
    //            PackageManager pm = context.getPackageManager();
    //
    //            ApplicationInfo applicationInfo = pm.getApplicationInfo(packageName, 0);
    //            PackageInfo packageInfo = pm.getPackageInfo(packageName, 0);
    //
    //            return new PackageMeta.Builder(applicationInfo.packageName)
    //                    .setLabel(applicationInfo.loadLabel(pm).toString())
    //                    .setHasSplits(applicationInfo.splitPublicSourceDirs != null && applicationInfo.splitPublicSourceDirs.length > 0)
    //                    .setIsSystemApp((applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0)
    //                    .setVersionCode(Utils.apiIsAtLeast(Build.VERSION_CODES.P) ? packageInfo.getLongVersionCode() : packageInfo.versionCode)
    //                    .setVersionName(packageInfo.versionName)
    //                    .setIcon(applicationInfo.icon)
    //                    .setInstallTime(packageInfo.firstInstallTime)
    //                    .setUpdateTime(packageInfo.lastUpdateTime)
    //                    .build();
    //
    //        } catch (PackageManager.NameNotFoundException e) {
    //            return null;
    //        }
    //    }

    protected constructor(`in`: Parcel) {
        packageName = `in`.readString() ?: ""
        label = `in`.readString()
        hasSplits = `in`.readByte().toInt() != 0
        isSystemApp = `in`.readByte().toInt() != 0
        versionCode = `in`.readLong()
        versionName = `in`.readString()
        iconUri = `in`.readParcelable(Uri::class.java.classLoader)
        installTime = `in`.readLong()
        updateTime = `in`.readLong()
        size = `in`.readString()
        sourceDir = `in`.readString()
        firstInstallTime = `in`.readLong()
        lastUpdateTime = `in`.readLong()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(packageName)
        dest.writeString(label)
        dest.writeByte((if (hasSplits) 1 else 0).toByte())
        dest.writeByte((if (isSystemApp) 1 else 0).toByte())
        dest.writeLong(versionCode)
        dest.writeString(versionName)
        dest.writeParcelable(iconUri, flags)
        dest.writeLong(installTime)
        dest.writeLong(updateTime)
        dest.writeString(size)
        dest.writeString(sourceDir)
        dest.writeLong(firstInstallTime)
        dest.writeLong(lastUpdateTime)
    }

    override fun describeContents(): Int {
        return 0
    }


    companion object CREATOR : Parcelable.Creator<PackageMeta> {
        override fun createFromParcel(parcel: Parcel): PackageMeta {
            return PackageMeta(parcel)
        }

        override fun newArray(size: Int): Array<PackageMeta?> {
            return arrayOfNulls(size)
        }
    }


}